File: AWS/iac-microbenchmark/lambdas/python/sha384.py
import boto3
import base64
import json
import os

# Initialize the boto3 client for KMS
kms_client = boto3.client('kms')

SIGN_ALGORITHM = 'HMAC_SHA_384'

def lambda_handler(event, context):

    # Extract the message from the event payload
    message = event.get('message')

    sha_kms_key_id = os.environ['SHA384_KMS_KEY_ARN']

    # Convert the message to bytes
    message_bytes = message.encode('utf-8')

    try:
        # Use KMS to sign (HMAC) the message
        response = kms_client.generate_mac(
            KeyId=sha_kms_key_id,
            Message=message_bytes,
            SigningAlgorithm=SIGN_ALGORITHM
        )
        # Base64 encode the signature for output
        signature = base64.b64encode(response['Mac']).decode('utf-8')
        return {
            'statusCode': 200,
        'headers' : {"Access-Control-Allow-Origin": "*",
                     "content-type": "application/json"},
            'body': json.dumps({
                'signature': signature
            })
        }
    except Exception as e:
        return {
            'statusCode': 500,
        'headers' : {"Access-Control-Allow-Origin": "*",
                     "content-type": "application/json"},
            'body': json.dumps({'error': str(e)})
        }

File: AWS/iac-microbenchmark/lambdas/typescript/sha384.ts
import { Context, APIGatewayProxyResult, APIGatewayEvent } from 'aws-lambda';
import { KMSClient, GenerateMacCommand } from '@aws-sdk/client-kms';
import * as process from 'process';

// Initialize the AWS KMS client
const kmsClient = new KMSClient({
  region: 'us-east-1' // Specify the desired region
});// Define the signing algorithm

const SIGN_ALGORITHM = 'HMAC_SHA_384';

export const handler = async (event: any, context: Context): Promise<any> => {

  // Extract the message from the event body, assuming it's passed as a JSON string
  const body = event.body ? JSON.parse(event.body) : {};
  const message: string = body.message || '';

  // Retrieve the KMS key ARN from environment variables
  const shaKmsKeyId: string = process.env.SHA384_KMS_KEY_ARN!;

  // Convert the message to a byte buffer
  const messageBytes = Buffer.from(message, 'utf-8');


  // Check if the KMS key ARN is provided
  if (!shaKmsKeyId) {
    throw new Error("SHA384_KMS_KEY_ARN environment variable is not set");

  }

  try {

    // Create the command to generate the HMAC
    const command = new GenerateMacCommand({
      KeyId: shaKmsKeyId,
      Message: messageBytes,
      MacAlgorithm: SIGN_ALGORITHM,
    });

    // Use KMS to generate HMAC of the message
    const response = await kmsClient.send(command);

    // Base64 encode the signature for output
    const signature = Buffer.from(response.Mac as Uint8Array).toString('base64');


    return {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      statusCode: 200,
      body: JSON.stringify({
        signature: signature
      }),
    };

  } catch (error) {
    console.error('Error signing message:', error);
    return {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      statusCode: 500,
      body: JSON.stringify({
        error: 'An error occurred while signing the message'
      }),
    };
  }
};
